### ISSUE TITLE
1_notmnist assignment1 zero mean equation in code

### ISSUE THREAD

**Issue by shadmanssk** [tensorflow/tensorflow#13002]
Okay in tutorial its say to make zero mean do r-128/128 but in code its r-128/255
Below is the code provided in ipython notebook
image_size = 28  # Pixel width and height.
pixel_depth = 255.0  # Number of levels per pixel.

def load_letter(folder, min_num_images):
  """Load the data for a single letter label."""
  image_files = os.listdir(folder)
  dataset = np.ndarray(shape=(len(image_files), image_size, image_size),
                         dtype=np.float32)
  print(folder)
  num_images = 0
  for image in image_files:
    image_file = os.path.join(folder, image)
    try:
      image_data = (ndimage.imread(image_file).astype(float) - 
                    pixel_depth / 2) / pixel_depth
      if image_data.shape != (image_size, image_size):
        raise Exception('Unexpected image shape: %s' % str(image_data.shape))
      dataset[num_images, :, :] = image_data
      num_images = num_images + 1
    except IOError as e:
      print('Could not read:', image_file, ':', e, '- it\'s ok, skipping.')

**Comment by cy89** [comment]
I'm sorry, @shadmanssk, but I don't understand what you're asking. Is there a bug or a feature request you'd like to make? Are you asking that we change or fix the tutorial? Or is there some aspect of the tutorial that you'd like explained? Also, which tutorial are you referring to?

**Comment by shadmanssk** [comment]
​
Tutorial: Deep Learning
Lesson 1 Concept 19 and Assignment 1

Assignment :
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/udacity/1_notmnist.ipynb

Issue:
Above equation of normalizing pixel inputs are not followed in code

Code (in above ipynb) :

image_size = 28  # Pixel width and height.pixel_depth = 255.0  #
Number of levels per pixel.


def load_letter(folder, min_num_images):
  """Load the data for a single letter label."""
  image_files = os.listdir(folder)
  dataset = np.ndarray(shape=(len(image_files), image_size, image_size),
                         dtype=np.float32)
  print(folder)
  num_images = 0
  for image in image_files:
    image_file = os.path.join(folder, image)
    try:
      image_data = (ndimage.imread(image_file).astype(float) -
                    pixel_depth / 2) / pixel_depth
      if image_data.shape != (image_size, image_size):
        raise Exception('Unexpected image shape: %s' % str(image_data.shape))

**Comment by cy89** [comment]
You're quite right that the Udacity slides and the code in the ipynb don't match up exactly. But I don't think it matters for educational purposes. 

I think there are two points in what Vincent is saying:
1. Having a zero mean makes it easier for the optimizer to work. 
2. Normalizing different axes to have the same scale (so circular, or standard deviations, rather than oval, as shown in slides) also makes it easier for the optimizer to work. 

I think the code gets both right. The -128 moves the mean of the [0,255] image data close to zero. 
The /255 rescales so that values will be in the range [-.5,.5]. The formulas that Vincent showed would have scaled to [-1,1]. But as long as everything rescales to the same ranges (so you get circularity across dimensions, not ovals), then the optimizer will still have an easier time. It's not the absolute magnitude of the rescaling that matters; it's the relative scales across the dimensions. 

Not sure if the course has gotten to batch normalization at this point, but that strives to rescale a standard deviation to have value about 1.0. Then you don't know the absolute max/min bounds, but you do have some statistical beliefs that your data are now spread out similarly in multiple dimensions. 

Does that make sense? @vincentvanhoucke, feel free to amend or comment, but I'll close for now.

**Comment by shadmanssk** [comment]
yeah even I realised that so I totally get your point still, it was
confusing at first but thanks for clearing out this thing.
